Per-user cartographer settings, split-screen hud fixes and more

Kantanomo opened the pull request:
Added a set of functions that will allow creating binary files that can be associated with any saved_game type (profile, variant, etc.)

- saved_games_async_helper_read_saved_game_bin
- saved_games_async_helper_write_saved_game_bin
- saved_games_async_helper_check_saved_game_bin

Created a new binary file that gets saved with the player profile cartographer_profile.bin, the file stores settings that were previously in the cartographer settings ini. 

The settings migrated are
- field_of_view
- static_first_person
- controller_sensitivity
- controller_modern
- controller_deadzone_type
- controller_custom_layout
- controller_deadzone (axial/radial)
- crosshair_offset
- crosshair_scale

The advanced settings menu has had some changes to make it support the multi-user use, it can be opened for the specific user on their controller using the back + d-pad up button combo. The menu has been changed so it limits the displayed settings to things that are specified in the cartographer profile. Global game settings have been limited to only being able to be modified by user 0.

hud rendering has been rewritten to re-implement the processes in the drawing function that do screen-split type hud changes (fullscreen, halfscreen, quarterscreen). The rewrite of these functions has allowed per-user crosshair scaling.

first person camera update has been re-written as well to add per-user crosshair_offset.

Kantanomo committed:
initial work

adds detours to increase the size of the player profile to store more data.

currently isn't setup to store anything or have any way to access stored data.

current issues
- on initial load the profile is blank on the profile select screen until you close and view it again.

Kantanomo committed:
more work

profiles now upgrade to whatever the new structure is.

rewrote a lot of loading related functions

fixed profiles not displaying right in the menus

Kantanomo committed:
misc work

-rewrote global initializers for saved game file memory
-added definition for c_player_profile_list
-rewrote c_player_profile_list::update_displayed_profiles
-lots of rewritten functions for saved_game operations
-rewrote warbird protected save_game_profile_load and saved_game_player_profile_read_file
-added input_abstraction_preferences_new

Kantanomo committed:
compile fix

Kantanomo committed:
creation of new bin file for cartographer per-profile settings

Kantanomo committed:
re-tooling and refactoring

re-tooled the code to read/write the cartographer bin files, to be more usable in the creation of any bin file for any valid saved game fileset.

refactored out the code for the cartographer_profile into its own file and set of functions for usage.

rewrote the player control fov hook to use the new per-profile settings.

Kantanomo committed:
load un-documented xinput function

load xinput function that allows us to read the home/guide button from controllers. eventually to be used to open advanced settings in that users context.

Kantanomo committed:
change advanced settings to be profile aware

advanced settings can now be set to modify a set profile.

moved fov and controller input settings to be loaded/stored in the new cartographer profile bin

Kantanomo committed:
home button usage and limited display

home buttons for opening the menu now works
limited the items that show in the menu for additional users to only display the per-profile settings
removing temp file

Kantanomo committed:
misc work

-initial work on rewriting hud_bitmap_widget draw (not complete special render cases not done)
-more settings moved to cartographer profile
--static first person
--crosshair scale
-added method to determine which screen split type a given user is playin in
-removed home button being used to open advanced settings due to conflicts with gamebar and steam
-new tag fixes function to fix split screen offsets on hud elements.

Berthalamew committed:
proper offset fix

Kantanomo committed:
shield meter bitmap widget rendering

finished the shield meter

Berthalamew committed:
add explanation

Berthalamew committed:
use original union for real_rectangle2d

Kantanomo committed:
territories bitmap widget

Kantanomo committed:
text widget work

incomplete doesn't actually work

Kantanomo committed:
text widgets completed and tag_fix fix

text widgets now rendering using the proper screen offset

fixed tag_fix for splitscreen offsets multiplying offsets multiple times due to tag_block space saving optimization in cache compiler.

Kantanomo committed:
rebase fixes

Kantanomo committed:
misc changes

-fixed a text_widget string formatting issue
-changed the controller combo for advanced settings to back+d_up
-fixed an issue with cartographer profiles not being signed out.

Kantanomo committed:
refactor and first_person_camera_update rewrite

refactored observer, camera, and director code into its proper places.

rewrote first_person_camera_update to allow per-profile crosshair offset.

Kantanomo committed:
first person fix and setting correction

Kantanomo committed:
some cleanup

Berthalamew commented on xlive/Blam/Engine/camera/first_person_camera.cpp:
```diff
@@ -0,0 +1,110 @@
+	observer_command->position.orientation = global_zero_vector3d;
+	observer_command->focus_distance = 0.f;
+	observer_command->forward = *unit_facing;
+	observer_command->field_of_view = 1.2217305f;
```
Supply the degrees instead of the radians value with the DEGREES_TO_RADIANS macro

Berthalamew commented on xlive/Blam/Engine/filesys/pc_saved_game.h:
```diff
@@ -0,0 +1,4 @@
+#pragma once
+
+#define k_max_save_game_fs_slot 0x8000
+#define save_game_max_name 0x80 
```
decimal values instead of hex

Berthalamew commented on xlive/H2MOD/Modules/Input/KeyboardInput.cpp:
```diff
@@ -214,6 +219,7 @@ void KeyboardInput::Initialize()
 	KeyboardInput::RegisterHotkey(&H2Config_hotkeyIdAlignWindow, hotkeyFuncAlignWindow);
 	KeyboardInput::RegisterHotkey(&H2Config_hotkeyIdWindowMode, hotkeyFuncWindowMode);
 	KeyboardInput::RegisterHotkey(&H2Config_hotkeyIdGuide, hotkeyFuncGuide);
+	KeyboardInput::RegisterHotkey(&pause, hotkeyFuncGuide2);
```
?

Kantanomo replied:
left over from testing, removed

Berthalamew commented on xlive/Blam/Engine/text/draw_string.cpp:
```diff
@@ -16,6 +16,11 @@ void __cdecl draw_string_set_shadow_color(real_argb_color const* color)
 	INVOKE(0x98A38, 0, draw_string_set_shadow_color, color);
 }
 
+void draw_string_set_unknown_color(real_argb_color const* color)
```
set_player_name_color

Berthalamew commented on xlive/Blam/Engine/saved_games/saved_game_files_async_windows.cpp:
```diff
@@ -0,0 +1,252 @@
+
+void saved_games_async_helpers_apply_hooks()
+{
+	//DETOUR_ATTACH(p_saved_games_async_helper_write_file, Memory::GetAddress<t_saved_games_async_helper_write_file>(0x42CA6), saved_games_async_helper_write_file);
```
?

Kantanomo replied:
Hooks are not needed but left code and hooks in case they need to be added in the future.

Berthalamew commented on xlive/Blam/Engine/saved_games/saved_game_files_async_windows.cpp:
```diff
@@ -0,0 +1,252 @@
+
+void saved_games_async_helper_read_file_success(int8* a1, s_file_reference* file_reference, int8* buffer, uint32 buffer_size)
+{
+	int32* unk_1 = ((int32*)a1 + 1);
```
implement the structure for this

Kantanomo replied:
it's an async related structure, i'll add a comment to replace with the structure when async is done.

Berthalamew commented on xlive/Blam/Engine/saved_games/saved_game_files_async_windows.cpp:
```diff
@@ -0,0 +1,252 @@
+				in_out_completion[2] = 1;
+				*(in_out_completion + 1) = 0;
+				*in_out_completion = 1;
+				*(in_out_completion + 514) = 0x3F800000;
```
implement the structure for this

Berthalamew commented on xlive/Blam/Engine/saved_games/saved_game_files.cpp:
```diff
@@ -0,0 +1,186 @@
+void saved_game_files_apply_hooks()
+{
+	WritePointer(Memory::GetAddress(0x39BD90), saved_game_files_memory_initialize);
+	//DETOUR_ATTACH(p_saved_games_load_save_file_information_from_disk, Memory::GetAddress<t_saved_games_load_save_file_information_from_disk>(0x46596), saved_games_load_save_file_information_from_disk);
```
?

Kantanomo replied:
commented out hook isn't needed.

Berthalamew commented on xlive/Blam/Engine/saved_games/player_profile.cpp:
```diff
@@ -0,0 +1,96 @@
+	if (result)
+		csmemcpy(profile, buffer, sizeof(s_saved_game_player_profile));
+
+	// todo: cleanup and define the structure here
```
complete todo

Kantanomo replied:
code isn't used in the project it's a partial rewrite of the warbird function, couldn't find anywhere where those properties were actually used in client. Will complete at a later time after PR?

Berthalamew commented on xlive/Blam/Engine/interface/new_hud_draw.cpp:
```diff
@@ -1,6 +1,738 @@
+
+/* globals */
+
+static const pixel32 g_draw_hud_bitmap_widget_shield_pixel_colors[9]
```
Create these using the D3DCOLOR macros

nukeulater commented on xlive/Blam/Engine/cseries/async.cpp:
```diff
@@ -1,8 +1,23 @@
+
+void __cdecl async_yield_until_done(int8* completion, bool play_sound)
+{
+	if (!completion[0])
```
with this condition here this is practically a while loop:

```cpp
while (!completion[0])
{
  ...
}
```

nukeulater commented on xlive/Blam/Engine/interface/user_interface_player_profile_list.cpp:
```diff
@@ -0,0 +1,156 @@
+		++current_child_widget_index;
+	}
+
+	if(profile_storage_size > 0)
```
same case here, while loop but for loop is preferable

nukeulater commented on xlive/Blam/Engine/interface/user_interface_player_profile_list.cpp:
```diff
@@ -0,0 +1,156 @@
+			enumerated_file_index = profile_storage[current_profile_storage_index].enumerated_file_index;
+
+			int32 current_player_profile_indices_index = 0;
+			while(true)
```
also a for loop

nukeulater commented on xlive/Blam/Engine/interface/user_interface_player_profile_list.cpp:
```diff
@@ -0,0 +1,156 @@
+			if(profile_storage_size > 0)
+			{
+				uint32 current_profile_storage_index = 0;
+				do
```
same here, for loop
de-compilation might represent this as a do while, since the code is optimized automatically by the compiler

nukeulater commented on xlive/Blam/Engine/saved_games/cartographer_player_profile.cpp:
```diff
@@ -0,0 +1,112 @@
+		datum user_datum = player_globals->player_user_mapping[user_index];
+		for(uint32 index = 0; index < k_number_of_controllers; index++)
+		{
+			if (player_globals->player_controller_mapping[index] == user_datum)
```
single if here, just the second condition on a new line

nukeulater commented on xlive/Blam/Engine/saved_games/cartographer_player_profile.h:
```diff
@@ -0,0 +1,55 @@
+
+struct s_saved_game_cartographer_player_profile_v1
+{
+	e_saved_game_cartographer_player_profile_version version;
```
a file signature wouldn't hurt along the version

nukeulater commented on xlive/Blam/Engine/saved_games/cartographer_player_profile.h:
```diff
@@ -0,0 +1,55 @@
+struct s_saved_game_cartographer_player_profile_v1
+{
+	e_saved_game_cartographer_player_profile_version version;
+	int32 field_of_view;
```
why int and not float?

Berthalamew replied:
Storing the value in degrees rather than radians?
If that's the case we can probably make this an uint8

nukeulater replied:
I would go as far as keeping this value in just 4 bits, obviously unsigned.

nukeulater replied:
idk, maybe a field of view of 78.1, or 78.2 degrees

Kantanomo replied:
data-type was just copied from what we used in the h2config

nukeulater replied:
no idea why this shouldn't just be a float

Kantanomo replied:
changed to real32

Kantanomo committed:
requested changes and profile upgrading

made requested changes for the pull request

implemented a upgrading process for the cartographer profile and added an explanation on how to add new ones.

Kantanomo committed:
requested change combine if statements

nukeulater commented on xlive/Blam/Engine/filesys/pc_saved_game.h:
```diff
@@ -1,4 +1,4 @@
-#define k_max_save_game_fs_slot 0x8000
-#define save_game_max_name 0x80 
\ No newline at end of file
+#define k_max_save_game_fs_slot SHRT_MAX
```
SHRT_MAX is 7FFF

Kantanomo replied:
fixed

nukeulater commented on xlive/Blam/Engine/interface/user_interface_player_profile_list.cpp:
```diff
@@ -50,52 +50,47 @@ void c_player_profile_list::update_displayed_profiles()
-	if(profile_storage_size > 0)
+	if (profile_storage_size > 0)
 	{
 		uint32 current_profile_storage_index = 0;
```
probably should've been removed, or don't redefine it in the for loop

Kantanomo replied:
fixed

Kantanomo committed:
fix define value and remove uneeded variables

Kantanomo committed:
requested changes and misc

changed player fov to real32

cleaned up profile upgrading code and added checks to make sure its not possible to get stuck in a loop.

added/updated some structures for director/cameras

Kantanomo committed:
crash fix and advanced settings fixes

fixed a crash when opening advanced settings from the cartographer menu in main menu

added a new function to test if a cartographer profile is signed in

made it so the advanced settings menu cannot be opened if a profile is not signed in.

Berthalamew committed:
cleanup and changes

- adjust cartographer player profile to use uint8's for fov
- remove unused config code
- hook up vehicle fov to the cartographer bins
- adjust position of functions and organize code into sections
- add assert in hud_widget_effect_evaluate
- remove incorrect switch break in c_cartographer_account_manager_edit_list
- remove implicit cast in rasterizer_dx9_sun_glow_draw

Berthalamew committed:
misc changes

- adjust async functions

Kantanomo committed:
sniper scope fix and casting changes.

Kantanomo committed:
ammo counter bitmap fix

this is disabled by default with comments pending a bitmap redraw, but the bitmaps were incorrectly resized during the port causing them to have extra padding that shouldn't be there. If the bitmaps aren't redone uncomment the if statement to have a general fix.
